src/messages/mod.rs

```diff
@@ -104,106 +135,136 @@ impl SignedRoutingMessage {
         self.src_section.as_ref()
     }
 
-    /// Returns whether the message is signed by the given public ID.
-    #[cfg(test)]
-    pub fn signed_by(&self, pub_id: &PublicId) -> bool {
-        self.signatures.contains_id(pub_id)
-    }
-
     /// Returns the number of nodes in the source authority.
     pub fn src_size(&self) -> usize {
         self.src_section.as_ref().map_or(0, |si| si.members().len())
     }
 
     /// Adds a proof if it is new, without validating it.
-    pub fn add_proof(&mut self, proof: Proof) {
-        if self.content.src.is_multiple() && self.is_sender(proof.pub_id()) {
-            let _ = self.signatures.add_proof(proof);
+    pub fn add_signature_share(
+        &mut self,
+        pk_share: BlsPublicKeyShare,
+        sig_share: BlsSignatureShare,
+    ) {
+        if let SecurityMetadata::Partial(ref mut sigs) = self.security_metadata {
+            let _ = sigs.insert(pk_share, sig_share);
         }
     }
 
-    /// Adds the given signature if it is new, without validating it. If the collection of section
-    /// lists isn't empty, the signature is only added if `pub_id` is a member of the first section
-    /// list.
-    #[cfg(test)]
-    pub fn add_signature(&mut self, pub_id: PublicId, sig: Signature) {
-        if self.content.src.is_multiple() && self.is_sender(&pub_id) {
-            let _ = self.signatures.sigs.insert(pub_id, sig);
+    /// Adds all signatures from the given message, without validating them.
+    pub fn add_signature_shares(&mut self, mut msg: SignedRoutingMessage) {
+        if self.content.src.is_multiple() {
+            if let (
+                SecurityMetadata::Partial(self_shares),
+                SecurityMetadata::Partial(other_shares),
+            ) = (&mut self.security_metadata, &mut msg.security_metadata)
+            {
+                self_shares.append(other_shares);
+            }
         }
     }
 
-    /// Adds all signatures from the given message, without validating them.
-    pub fn add_signatures(&mut self, msg: SignedRoutingMessage) {
-        if self.content.src.is_multiple() {
-            self.signatures.merge(msg.signatures);
+    /// Combines the signatures into a single BLS signature
+    pub fn combine_signatures(&mut self, pk_set: &BlsPublicKeySet) {
+        if let SecurityMetadata::Partial(sigs) =
+            mem::replace(&mut self.security_metadata, SecurityMetadata::None)
+        {
+            if let Some(full_sig) =
+                pk_set.combine_signatures(sigs.iter().map(|(key, sig)| (*key, sig)))
+            {
+                self.security_metadata = SecurityMetadata::Signed(full_sig);
+            } else {
+                log_or_panic!(
+                    LogLevel::Error,
+                    "Combining signatures failed on {:?}!",
+                    self
+                );
+            }
+        } else {
+            warn!("Tried to call combine_signatures on {:?}", self);
```

Comment:
log_or_panic?

Reply:
I considered it, but it's not a good idea. We can sometimes call this function on a message with `SecurityMetadata::None`. I could do a `match` and only `log_or_panic` on the 100% incorrect variants, maybe?

Reply:
If it can happen in normal case then the warning will be spammy will it not?
Should we not warn then?

Reply:
Right, good point. I'll change this.